driver/vhci/usbreq.c

```diff
@@ -104,8 +104,8 @@ static void
 remove_cancelled_urbr(pusbip_vpdo_dev_t vpdo, PIRP irp)
 {
 	struct urb_req	*urbr;
-
-	KeAcquireSpinLockAtDpcLevel(&vpdo->lock_urbr);
```

Comment:
The previous code would suffice. Because [MSDN](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-iosetcancelroutine) says
> Driver Cancel routines are called at IRQL = DISPATCH_LEVEL with the cancel spin lock held.

Reply:
i'm releasing early CancelSpinlock as we don't need it, this is for os responsivnes.

From [MSDN](https://docs.microsoft.com/pl-pl/windows-hardware/drivers/ddi/wdm/nc-wdm-driver_cancel):
> The I/O manager calls IoAcquireCancelSpinLock before calling a driver's Cancel routine, so the Cancel routine must call IoReleaseCancelSpinLock at some point. The routine should not hold the spin lock longer than necessary.

So to be nice for the rest of drivers we release it asap - multicore.

Reply:
`KeAcuiqreSpinLockAtDpcLevel` may be faster than `KeAcquireSpinLock()`. My comment meant that `KeAcquireSpinLockAtDpcLevel` should be used instead  of `KeAcquireSpinLock()` because `remove_cancelled_urbr` is already at DISPATCH_LEVEL.

Reply:
yeah, i'm aware, but with this change it is no more at the dpc level at this point in a code

Reply:
Which routine calls`remove_cancelled_urbr` with a lower IRQL than DISPATCH?

Reply:
it's not much about calling irql lower than dispatch, it's also about locking order
we have 2 possible locks:
A) lock_urbr
B) global cancel lock

lets say we have functions acquire - a, release - r

Throughout the code we call
a(A) -> a(B) -> (...) -> r(B) -> r(A)

so actually in the cancel routine if we were holding lock B and acquiring A we would be doing
(from cancel routine caller) a(B) -> a(A) -> r(A) -> r(B)

this might have resulted in a deadlock if we had on different processor A locked,
so we are doing in this patchset:
(from cancel routine caller) a(B) -> r(B) -> a(A) -> r(A)

removing the potential deadlock :)

Reply:
I'm just saying once again that `KeAcquireSpinLockAtDpcLevel` should be used instead of `KeAcquireSpinLock()` because `remove_cancelled_urbr` is already at DISPATCH_LEVEL. 😢 Please remind that our long discussion is confined to code line 107-108. Of course, your deadlock comment should be considered later to review below codes.

Reply:
this method is called in this patchset at usbreq.c:142:
```
	IoReleaseCancelSpinLock(irp->CancelIrql);

	remove_cancelled_urbr(vpdo, irp);
```
we are releasing cancel spinlock and we are no longer at passive dispatch_level so this is okay :)

Reply:
I got it finally.. :smiley: Sorry for long discussion.